streams/src/main/java/org/apache/kafka/streams/state/internals/buffer/InMemoryTimeOrderedKeyValueBufferStore.java

```diff
@@ -0,0 +1,219 @@
+ */
+package org.apache.kafka.streams.state.internals.buffer;
+
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.common.header.Header;
+import org.apache.kafka.common.header.Headers;
+import org.apache.kafka.common.header.internals.RecordHeaders;
+import org.apache.kafka.common.serialization.ByteArraySerializer;
+import org.apache.kafka.common.serialization.BytesSerializer;
+import org.apache.kafka.common.utils.Bytes;
+import org.apache.kafka.streams.KeyValue;
+import org.apache.kafka.streams.processor.ProcessorContext;
+import org.apache.kafka.streams.processor.StateStore;
+import org.apache.kafka.streams.processor.internals.ProcessorRecordContext;
+import org.apache.kafka.streams.processor.internals.ProcessorStateManager;
+import org.apache.kafka.streams.processor.internals.RecordBatchingStateRestoreCallback;
+import org.apache.kafka.streams.processor.internals.RecordCollector;
+import org.apache.kafka.streams.state.StoreBuilder;
+import org.apache.kafka.streams.state.internals.ContextualRecord;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.nio.ByteBuffer;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.LinkedList;
+import java.util.Map;
+
+public class InMemoryTimeOrderedKeyValueBufferStore implements StateStore {
+    private static final Logger LOG = LoggerFactory.getLogger(InMemoryTimeOrderedKeyValueBufferStore.class);
+    private static final String BUFFER_TIME_HEADER = "_bufferTime";
+    private static final BytesSerializer KEY_SERIALIZER = new BytesSerializer();
+    private static final ByteArraySerializer VALUE_SERIALIZER = new ByteArraySerializer();
+
+    private final InMemoryTimeOrderedKeyValueBuffer buffer;
+    private final String storeName;
+
+    private RecordCollector collector;
+    private String changelogTopic;
+    private boolean open;
+
+    public static class Builder implements StoreBuilder<StateStore> {
+        private final String storeName;
+
+        public Builder(final String storeName) {
+            this.storeName = storeName;
+        }
+
+        @Override
+        public StoreBuilder<StateStore> withCachingEnabled() {
+            return null;
+        }
+
+        @Override
+        public StoreBuilder<StateStore> withCachingDisabled() {
+            return null;
+        }
+
+        @Override
+        public StoreBuilder<StateStore> withLoggingEnabled(final Map<String, String> config) {
+            return null;
+        }
+
+        @Override
+        public StoreBuilder<StateStore> withLoggingDisabled() {
+            return null;
+        }
+
+        @Override
+        public StateStore build() {
+            return new InMemoryTimeOrderedKeyValueBufferStore(
+                new InMemoryTimeOrderedKeyValueBuffer(),
+                storeName
+            );
+        }
+
+        @Override
+        public Map<String, String> logConfig() {
+            return Collections.emptyMap();
```

Comment:
Is there anything I should consider here?

Reply:
Not for now IMHO -- no public `BufferStoreInterface` yet.